### PR TITLE
Mesh-605 RPC call to read reward curve params

### PR DESCRIPTION
**Pull request by vkandy**
Added `pallet-staking-rpc` and `pallet-staking-rpc-runtime-api` to call into staking runtime. 

**RPC Request** 
`
curl -H 'Content-Type: application/json' --data '{ "jsonrpc":"2.0", "method":"staking_getCurve", "id": 1 }'
`

**Curve 1**
Parameters
```
        min_inflation: 0_500_000,
        max_inflation: 1_000_000,
        ideal_stake: 0_100_000,
        falloff: 0_050_000,
        max_piece_count: 40,
        test_precision: 0_005_000,
```

Output of `get_curve` RPC call:
`
{"jsonrpc":"2.0","result":[[0,500000000],[100000000,1000000000],[105374000,964103000],[111181000,928206000],[117496000,892309000],[124419000,856412000],[132076000,820515000],[140644000,784618000],[150369000,748721000],[161613000,712824000],[174938000,676927000],[191296000,641030000],[212485000,605133000],[242616000,569236000],[278513000,542093000],[314410000,525591000],[350305000,515559000],[386197000,509460000],[422085000,505752000],[457962000,503498000],[493848000,502127000],[529697000,501294000],[565567000,500787000],[601384000,500479000],[637087000,500292000],[672791000,500178000],[708169000,500109000],[743273000,500067000],[778700000,500041000],[814385000,500025000],[846578000,500016000],[880481000,500010000],[906210000,500007000],[930481000,500005000],[946578000,500004000],[967330000,500003000],[996578000,500002000],[1000000000,500000000]],"id":1}
`

**Curve 2 (big falloff)**

Parameters:
```
            min_inflation: 0_100_000,
            max_inflation: 0_400_000,
            ideal_stake: 0_400_000,
            falloff: 1_000_000,
            max_piece_count: 40,
            test_precision: 0_005_000,
```

Output of `get_curve` RPC call:

`{"jsonrpc":"2.0","result":[[0,100000000],[400000000,400000000],[423073000,395240000],[446145000,390556000],[469219000,385946000],[492293000,381409000],[515367000,376944000],[538440000,372550000],[561512000,368226000],[584587000,363970000],[607660000,359782000],[630734000,355660000],[653806000,351604000],[676880000,347612000],[699954000,343683000],[723026000,339817000],[746100000,336012000],[769176000,332267000],[792249000,328582000],[815324000,324955000],[838395000,321386000],[861466000,317874000],[884541000,314417000],[907615000,311015000],[930688000,307667000],[953763000,304372000],[976839000,301129000],[999912000,297938000],[1000000000,297926000]],"id":1}`

### CODE REVIEW

File: pallets/staking/src/lib.rs [hunk 1]
```diff
@@ -1998,6 +1998,16 @@ impl<T: Trait> Module<T> {
         true
     }
 
+    /// Return reward curve points
+    pub fn get_curve() -> Vec<(Perbill, Perbill)> {
+        let curve = &T::RewardCurve::get();
+        let mut points: Vec<(Perbill, Perbill)> = Vec::new();
```
Comment (satyamakgec) [inline pallets/staking/src/lib.rs]: You can use `with_capacity()` for the vector to determine the exact length

File: pallets/staking/rpc/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+impl<C, Block> StakingApi<<Block as BlockT>::Hash> for Staking<C, Block>
+where
+    Block: BlockT,
+    C: Send + Sync + 'static,
+    C: ProvideRuntimeApi,
+    C: HeaderBackend<Block>,
+    C::Api: StakingRuntimeApi<Block>,
+{
+    fn get_curve(&self, at: Option<<Block as BlockT>::Hash>) -> Result<Vec<(Perbill, Perbill)>> {
```
Comment (satyamakgec) [inline pallets/staking/rpc/src/lib.rs]: @vkandy If the RewardCurve points are fixed i.e (Unchangeable as the block grows) then you can remove the block hash part.

File: pallets/staking/src/lib.rs (hunk 1 above)
Comment (vkandy) [reply pallets/staking/src/lib.rs]: The number of points change depending on the parameters (as in the description). So we don't know the capacity in advance to be able to use `Vec::with_capacity()`.

File: pallets/staking/rpc/src/lib.rs (hunk 2 above)
Comment (vkandy) [reply pallets/staking/rpc/src/lib.rs]: The block is required to query runtime _as of_ a certain block.  Currently the RPC assumes the best block.

File: pallets/staking/src/lib.rs (hunk 1 above)
Comment (satyamakgec) [reply pallets/staking/src/lib.rs]: can't we have Vec::with_capacity(curve.points.len())

File: pallets/staking/rpc/src/lib.rs (hunk 2 above)
Comment (satyamakgec) [reply pallets/staking/rpc/src/lib.rs]: We can change that if block is not needed, I think the authority module has some RPC without block hash.

**Pull request merged**